File: 4 Kyu/4 Kyu - Large Factorials.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_DIGIT 1000

char *factorial(int n) {

  if (n < 0)
      return "";

  size_t x, i, tmp;
  size_t next = 0, carry = 0, ndigits = 1, buffer[MAX_DIGIT] = {1};

  for (x = 2; x <= n; x++) {

      for (i = 0; i < ndigits; ++i) {
          tmp = buffer[i] * x;
          buffer[i] = (tmp + carry) % 10;
          carry = (tmp + carry) / 10;

          if (i == ndigits - 1 && carry > 0)
                ndigits++;
      }

}

  char *result = calloc ((ndigits + 1),sizeof (char));

  while (ndigits-->0)
        result[next++] = buffer[ndigits] + '0';

  return result;
}

File: 4 Kyu/4 Kyu - Large Factorials.rs
fn factorial (n:u32) -> String {
    let mut fact = String::new();
    let mut carry = 0;
    let mut ndigit = 1;
    let mut buffer = [0;1024];
    buffer[0] = 1;

    for x in 2..=n {
        let mut i = 0;

        while i < ndigit {
            let tmp = buffer[i] * x;
            buffer[i] = (tmp + carry) % 10;
            carry = (tmp + carry) / 10;

            if i == (ndigit - 1) && (carry > 0) {
                ndigit += 1;
            }

            i += 1;
        }
    }

    while ndigit > 0 {
        ndigit -= 1;
        fact += &format! ("{}", buffer[ndigit]);
    }

    fact
}

fn main () {

}
